### PR TITLE
Add in support for reloading and restarting all running apps

### PR DESCRIPTION
**Pull request by DaveShuckerow**
You can:

* use alt/option as a modifier on the normal ctrl/cmd + \ 
* click the 'reload all' or 'restart all' button

Not done:
* [ ] Differentiate the all icons from the singe icons.  I think the best way to do this would be to add the number of running apps to the button, just like the 'run' button at the top of the screen does.  I'm not sure how to do this.
* [ ] Decide if we want to have the buttons, or if it's better to just use the key shortcuts.  This is a fairly esoteric use case.

### CODE REVIEW

**Comment by devoncarew** [comment]
Can you post a screen shot (just cropped to the icons is fine).

Another way to implement this is to have a preference in the settings for reload on save (and the toolbar reload actions) to imply 'reload across all devices'. The reload actions specific to the run and debug tool windows would stay specific to each app instance.

**Comment by devoncarew** [comment]
(my hesitation here is w/ adding another top-level toolbar icon - it's already pretty busy up there, and our bar to adding new ones is likely that it'll be frequently used by a large set of our users)

**Comment by DaveShuckerow** [comment]
I think, since this is a niche use case, I'll pull out the icons and leave the keyboard shortcut.

I set this up to use the alt/option modifier key when doing a hot reload to perform a reload-all.  I think that will be more versatile than having to check a settings box to switch between reload and reload-all.

**Comment by DaveShuckerow** [comment]
PTAL, let me know if you have additional questions or comments.

**Comment by devoncarew** [comment]
Looking now, sorry for the delay.

Can you post a screenshot of the Run menu (it looks like you're contributing the UI actions there).

**Comment by devoncarew** [comment]
The changes look good, I just want to make sure I understand the UI changes:

- new contributions to the run menu
- new actions available through keybindings

Am I missing anything?

File: resources/META-INF/plugin.xml [hunk 1]
```diff
@@ -730,6 +730,18 @@
...
         <keyboard-shortcut keymap="$default" first-keystroke="ctrl shift S"/>
       </action>
+      <action id="Flutter.Toolbar.ReloadAllAction" class="io.flutter.actions.ReloadAllFlutterAppsRetarget"
+              description="Reload All"
+              icon="FlutterIcons.HotReload">
+        <keyboard-shortcut keymap="$default" first-keystroke="ctrl alt BACK_SLASH"/>
+        <keyboard-shortcut keymap="$default" first-keystroke="ctrl alt S"/>
+      </action>
+      <action id="Flutter.Toolbar.RestartAllAction" class="io.flutter.actions.RestartAllFlutterAppsRetarget"
+              description="Restart All"
```
[the remark this answers is not in the document]
Comment (DaveShuckerow) [reply resources/META-INF/plugin.xml]: Done

File: resources/META-INF/plugin.xml [hunk 2]
```diff
@@ -730,6 +730,18 @@
         <keyboard-shortcut keymap="$default" first-keystroke="ctrl shift BACK_SLASH"/>
         <keyboard-shortcut keymap="$default" first-keystroke="ctrl shift S"/>
       </action>
+      <action id="Flutter.Toolbar.ReloadAllAction" class="io.flutter.actions.ReloadAllFlutterAppsRetarget"
+              description="Reload All"
```
[the remark this answers is not in the document]
Comment (DaveShuckerow) [reply resources/META-INF/plugin.xml]: Good idea; done.

File: src/io/flutter/actions/ReloadAllFlutterApps.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2016 The Chromium Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (DaveShuckerow) [reply src/io/flutter/actions/ReloadAllFlutterApps.java]: Done

File: src/io/flutter/actions/RestartAllFlutterApps.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2016 The Chromium Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (DaveShuckerow) [reply src/io/flutter/actions/RestartAllFlutterApps.java]: Done

File: src/io/flutter/console/FlutterConsoleFilter.java [hunk 5]
```diff
@@ -189,7 +189,7 @@ public void navigate(final Project project) {
         return;
       }
 
-      final FlutterApp app = FlutterApp.fromProjectProcess(project);
+      final FlutterApp app = FlutterApp.firstFromProjectProcess(project);
```
[the remark this answers is not in the document]
Comment (DaveShuckerow) [reply src/io/flutter/console/FlutterConsoleFilter.java]: 👍

File: src/io/flutter/run/FlutterReloadManager.java [hunk 6]
```diff
@@ -251,9 +249,21 @@ else if (result.isRestartRecommended()) {
     }
   }
 
-  public void saveAllAndRestart(@NotNull FlutterApp app, String reason) {
+  public void saveAllAndReload(@NotNull FlutterApp app, @NotNull String reason) {
```
[the remark this answers is not in the document]
Comment (DaveShuckerow) [reply src/io/flutter/run/FlutterReloadManager.java]: saveAllAndReload is the old name -- it refers to saving all the open files in contrast with only saving the file currently being edited.

File: src/io/flutter/run/FlutterReloadManager.java [hunk 7]
```diff
@@ -251,9 +249,21 @@ else if (result.isRestartRecommended()) {
...
+    FileDocumentManager.getInstance().saveAllDocuments();
+    reloadApp(app, reason);
+  }
+
+  public void saveAllAndReloadAll(@NotNull ImmutableList<FlutterApp> appsToReload, @NotNull String reason) {
```
[the remark this answers is not in the document]
Comment (DaveShuckerow) [reply src/io/flutter/run/FlutterReloadManager.java]: ditto

[the remark this answers is not in the document]
Comment (DaveShuckerow) [reply src/io/flutter/run/FlutterReloadManager.java]: Done, both here and above.

File: src/io/flutter/run/daemon/FlutterApp.java [hunk 8]
```diff
@@ -209,6 +210,24 @@ public static FlutterApp fromProjectProcess(@NotNull Project project) {
     return null;
   }
 
+  @NotNull
+  public static ImmutableList<FlutterApp> allFromProjectProcess(@NotNull Project project) {
```
[the remark this answers is not in the document]
Comment (DaveShuckerow) [reply src/io/flutter/run/daemon/FlutterApp.java]: Done

**Comment by DaveShuckerow** [comment]
> The changes look good, I just want to make sure I understand the UI changes:
> 
> * new contributions to the run menu
> * new actions available through keybindings
> 
> Am I missing anything?

From what I can tell, these changes don't actually show up in the run menu.  What I did appears to be the minimum necessary to enable just the keyboard shortcuts.

**Comment by devoncarew** [comment]
> From what I can tell, these changes don't actually show up in the run menu.

Very odd, as it looks like they're defined in the run menu actions section.

Screen Shot 2019-03-19 at 8 35 43 AM

Can you do a bit more investigation here (and essentially convince yourself that this is the right place to contribute these actions)? Thanks!

**Comment by DaveShuckerow** [comment]
Interesting: the reload actions are all visible in the run menu when running an app via the external tools, but not when running an app via the g3 tools:

Screenshot from 2019-03-19 13-54-36

I don't know what would cause that.

**Comment by DaveShuckerow** [comment]
Since that doesn't affect the utility of this change, I've filed it in a separate internal issue (at 128927362) and will merge this.

**Pull request merged**